storm-core/src/jvm/backtype/storm/scheduler/resource/strategies/ResourceAwareStrategy.java

```diff
@@ -0,0 +1,478 @@
+                    neighbors.addAll(comp.parents);
+                    for (String nbID : neighbors) {
+                        if (!visited.containsKey(nbID)) {
+                            Component child = topologies.getAllComponents().get(td.getId()).get(nbID);
+                            queue.offer(child);
+                        }
+                    }
+                }
+            }
+        }
+        return ordered__Component_list;
+    }
+
+    private List<Component> getSpouts(Topologies topologies, TopologyDetails td) {
+        List<Component> spouts = new ArrayList<Component>();
+        for (Component c : topologies.getAllComponents().get(td.getId())
+                .values()) {
+            if (c.type == Component.ComponentType.SPOUT) {
+                spouts.add(c);
+            }
+        }
+        return spouts;
+    }
+
+    private Integer getLongestPriorityListSize(Map<Integer, List<ExecutorDetails>> priorityToExecutorMap) {
+        Integer mostNum = 0;
+        for (List<ExecutorDetails> execs : priorityToExecutorMap.values()) {
+            Integer numExecs = execs.size();
+            if (mostNum < numExecs) {
+                mostNum = numExecs;
+            }
+        }
+        return mostNum;
+    }
+
+    /**
+     * Get the remaining amount memory that can be assigned to a worker given the set worker max heap size
+     * @param ws
+     * @param td
+     * @param scheduleAssignmentMap
+     * @return The remaining amount of memory
+     */
+    private Double getWorkerScheduledMemoryAvailable(WorkerSlot ws, TopologyDetails td, Map<WorkerSlot, Collection<ExecutorDetails>> scheduleAssignmentMap) {
+        Double memScheduleUsed = this.getWorkerScheduledMemoryUse(ws, td, scheduleAssignmentMap);
+        return td.getTopologyWorkerMaxHeapSize() - memScheduleUsed;
+    }
+
+    /**
+     * Get the amount of memory already assigned to a worker
+     * @param ws
+     * @param td
+     * @param scheduleAssignmentMap
+     * @return the amount of memory
+     */
+    private Double getWorkerScheduledMemoryUse(WorkerSlot ws, TopologyDetails td, Map<WorkerSlot, Collection<ExecutorDetails>> scheduleAssignmentMap) {
+        Double totalMem = 0.0;
+        Collection<ExecutorDetails> execs = scheduleAssignmentMap.get(ws);
+        if(execs != null) {
+            for(ExecutorDetails exec : execs) {
+                totalMem += td.getTotalMemReqTask(exec);
+            }
+        } 
+        return totalMem;
+    }
+
+    /**
+     * Checks whether we can schedule an Executor exec on the worker slot ws
+     * @param exec
+     * @param ws
+     * @param td
+     * @param scheduleAssignmentMap
+     * @return a boolean: True denoting the exec can be scheduled on ws and false if it cannot
+     */
+    private boolean checkWorkerConstraints(ExecutorDetails exec, WorkerSlot ws, TopologyDetails td, Map<WorkerSlot, Collection<ExecutorDetails>> scheduleAssignmentMap) {
```

Comment:
Since we now only consider the memory when checkWorkerConstraints, may add some comments saying that we will also check CPU constraint in the near future.